core/src/main/java/io/questdb/cairo/ColumnVersionWriter.java

```diff
@@ -88,20 +89,58 @@ public boolean hasChanges() {
     public void removeColumnTop(long partitionTimestamp, int columnIndex) {
         int recordIndex = getRecordIndex(partitionTimestamp, columnIndex);
         if (recordIndex >= 0) {
-            getCachedList().setQuick(recordIndex + 3, 0);
+            cachedList.setQuick(recordIndex + COLUMN_TOP_OFFSET, 0);
+            hasChanges = true;
+        }
+    }
+
+    public void detachPartitionColumns(long partitionTimestamp) {
+        int from = cachedList.binarySearchBlock(BLOCK_SIZE_MSB, partitionTimestamp, BinarySearch.SCAN_UP);
+        if (from > -1) {
+            int to = cachedList.binarySearchBlock(from, BLOCK_SIZE_MSB, partitionTimestamp, BinarySearch.SCAN_DOWN);
+            int len = to - from + BLOCK_SIZE;
+            cachedList.removeIndexBlock(from, len);
             hasChanges = true;
         }
     }
 
     public void truncate() {
-        LongList cachedList = getCachedList();
         if (cachedList.size() > 0) {
             cachedList.clear();
             hasChanges = true;
             commit();
         }
     }
 
+    public void upsertPartition(long partitionTimestamp, ColumnVersionReader source) {
+        LongList src = source.cachedList;
+        LongList dest = this.cachedList;
+
+        int srcIndex = src.binarySearchBlock(BLOCK_SIZE_MSB, partitionTimestamp, BinarySearch.SCAN_UP);
+        if (srcIndex < 0) { // source does not have partition information
+            return;
+        }
+
+        int index = dest.binarySearchBlock(BLOCK_SIZE_MSB, partitionTimestamp, BinarySearch.SCAN_UP);
+        if (index > -1L) {
+            // Wipe out all the information about this partition to replace with the new one.
+            detachPartitionColumns(partitionTimestamp);
+            index = dest.binarySearchBlock(BLOCK_SIZE_MSB, partitionTimestamp, BinarySearch.SCAN_UP);
+        }
+
+        if (index < 0) { // the cache does not contain this partition
+            index = -index - 1;
+            int srcEnd = src.binarySearchBlock(srcIndex, BLOCK_SIZE_MSB, partitionTimestamp, BinarySearch.SCAN_DOWN);
+            dest.insertFromSource(index, src, srcIndex, srcEnd + BLOCK_SIZE);
+        } else {
+            throw CairoException.instance(0)
```

Comment:
this is confusing and exception message does not help. "invalid" partition in what way? Can we be more specific with the error?

If we expect partition not to exist in destination, why don't we check it right away and state our expectation?

Reply:
It can happen when data in ColumnVersionWriter is not sorted correctly for example e.g. the state is bombed.
I changed message to "invalid Column Version state".